rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+want access to all the intermediate tensor values computed in a FuncGraph.
+
+### Design: Debug Callbacks for Op Instrumentation
+
+The following API meets the requirements listed above. The code below shows the
+API and the detailed signature and semantics of the callbacks that can be passed
+to the API.
+
+```python
+# Exported publicly as: tf.debugging.op_callback()
+def op_callback(callback_fn):
+  """Intercepts op execution and op creation.
+
+  The `callback_fn` will be invoked immediately after any of the three types
+  of events:
+    - The execution of an op under eager mode,
+    - The execution of a FuncGraph under eager mode,
+    - The creation of an op during graph construction (e.g., in
+      @tf.function-decorated Python functions.)
+
+  Args:
+    A callback_fn that has the following signature:
+      def callback_fn(op_type,
+                      inputs,
+                      attrs,
+                      outputs,
+                      op_name=None,
+                      graph=None):
+        # op_type: The type of the op. E.g., "MatMul".
+        #          For the special case of FuncGraph execution, op_type
+        #          takes the special value `tf.Graph`.
+        # inputs: (`tuple` of `Tensor`s) Input tensors to the op or the
+        #         FuncGraph.
+        #         In eager execution, these are `EagerTensor`s.
+        #         In graph construction, these are non-eager `Tensor`s
+        #         that form the inputs to the just-created op.
+        # attrs: The attributes of the op or FuncGraph of which the execution
+        #        or creation caused the current invocation of the callback.
+        # outputs: (`tuple of `Tensor`s) Output tensors from the op or
+        #          FuncGraph.
+        #          In eager execution, these are `EagerTensor`s.
+        #          In graph construction, these are non-eager `Tensor`s that
+        #          are the outputs of the just-created op.
+        # op_name: Name of the op or FuncGraph.
+        #          If the current invocation of the callback is due to the
+        #          eager execution of an op, this will be `None` as op names
+        #          are meaningless in eager execution.
+        #          If this invocation of the callback is due to the
+        #          eager execution of a FuncGraph, this will be the
+        #          internally-generated name of the FuncGraph.
+        #          In graph construction, this is the name of the op.
+        # graph: The graph that the op belongs to (if any).
+        #        In eager execution of an op, this is `None`.
+        #        In eager execution of a FuncGraph, the is the FuncGraph
+        #        object per se.
+        #        In graph construction, this is the op's containing graph.
+        #
+        # Return values:
+        #   This callback function is expected to return a `list` or `tuple`
```

Comment:
Presumably in the same order? Is there value in allowing different return types? Eg, nested things?